src/libsyntax/parse/parser.rs

```diff
@@ -637,16 +642,28 @@ impl<'a> Parser<'a> {
             } else {
```

Comment:
This branch seems like something to refactor out to one or more functions/methods in `diagnostics.rs`?